File: services/api/src/modules/trips/trips.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Query } from '@nestjs/common';
import { ApiHeader, ApiTags } from '@nestjs/swagger';
import { CurrentUserId } from '../../common/decorators/current-user-id.decorator';
import { PaginationQueryDto } from '../../common/dto/pagination-query.dto';
import { AutoAssembleTripDto } from './dto/auto-assemble-trip.dto';
import { CreateTripPointDto } from './dto/create-trip-point.dto';
import { CreateTripDto } from './dto/create-trip.dto';
import { PublishTripDto } from './dto/publish-trip.dto';
import { ReorderTripPointsDto } from './dto/reorder-trip-points.dto';
import { UpdateTripPointDto } from './dto/update-trip-point.dto';
import { UpdateTripDto } from './dto/update-trip.dto';
import { TripsService } from './trips.service';

@ApiTags('trips')
@ApiHeader({
  name: 'x-user-id',
  required: false,
  description: 'Temporary dev auth header. Falls back to DEMO_USER_ID.',
})
@Controller('v1/trips')
export class TripsController {
  constructor(private readonly tripsService: TripsService) {}

  @Post()
  create(@CurrentUserId() userId: string, @Body() dto: CreateTripDto) {
    return this.tripsService.createTrip(userId, dto);
  }

  @Get()
  list(@CurrentUserId() userId: string, @Query() query: PaginationQueryDto) {
    return this.tripsService.listTrips(userId, query);
  }

  @Get(':tripId')
  getOne(@CurrentUserId() userId: string, @Param('tripId') tripId: string) {
    return this.tripsService.getTrip(userId, tripId);
  }

  @Patch(':tripId')
  update(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Body() dto: UpdateTripDto,
  ) {
    return this.tripsService.updateTrip(userId, tripId, dto);
  }

  @Post(':tripId/points')
  createPoint(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Body() dto: CreateTripPointDto,
  ) {
    return this.tripsService.createPoint(userId, tripId, dto);
  }

  @Patch(':tripId/points/:pointId')
  updatePoint(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Param('pointId') pointId: string,
    @Body() dto: UpdateTripPointDto,
  ) {
    return this.tripsService.updatePoint(userId, tripId, pointId, dto);
  }

  @Delete(':tripId/points/:pointId')
  deletePoint(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Param('pointId') pointId: string,
  ) {
    return this.tripsService.deletePoint(userId, tripId, pointId);
  }

  @Post(':tripId/points/reorder')
  reorderPoints(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Body() dto: ReorderTripPointsDto,
  ) {
    return this.tripsService.reorderPoints(userId, tripId, dto);
  }

  @Post(':tripId/auto-assemble')
  autoAssemble(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Body() dto: AutoAssembleTripDto,
  ) {
    return this.tripsService.autoAssembleTrip(userId, tripId, dto);
  }

  @Post(':tripId/publish')
  publish(
    @CurrentUserId() userId: string,
    @Param('tripId') tripId: string,
    @Body() dto: PublishTripDto,
  ) {
    return this.tripsService.publishTrip(userId, tripId, dto);
  }
}

File: services/api-java/src/main/java/com/tripin/api/controller/TripsController.java
package com.tripin.api.controller;

import com.tripin.api.service.TripsService;
import com.tripin.api.web.CurrentUserResolver;
import com.tripin.api.web.Requests.AutoAssembleTripRequest;
import com.tripin.api.web.Requests.CreateTripPointRequest;
import com.tripin.api.web.Requests.CreateTripRequest;
import com.tripin.api.web.Requests.PublishTripRequest;
import com.tripin.api.web.Requests.ReorderTripPointsRequest;
import java.util.Map;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/v1/trips")
public class TripsController {
  private final TripsService tripsService;
  private final CurrentUserResolver currentUserResolver;

  public TripsController(TripsService tripsService, CurrentUserResolver currentUserResolver) {
    this.tripsService = tripsService;
    this.currentUserResolver = currentUserResolver;
  }

  @PostMapping
  public Map<String, Object> create(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @RequestBody CreateTripRequest request) {
    return tripsService.createTrip(currentUserResolver.resolve(userId), request);
  }

  @GetMapping
  public Map<String, Object> list(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @RequestParam(required = false) Integer limit,
      @RequestParam(required = false) String cursor) {
    return tripsService.listTrips(currentUserResolver.resolve(userId), limit, cursor);
  }

  @GetMapping("/{tripId}")
  public Map<String, Object> getOne(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId) {
    return tripsService.getTrip(currentUserResolver.resolve(userId), tripId);
  }

  @PatchMapping("/{tripId}")
  public Map<String, Object> update(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @RequestBody(required = false) CreateTripRequest request) {
    return tripsService.updateTrip(currentUserResolver.resolve(userId), tripId, request);
  }

  @PostMapping("/{tripId}/points")
  public Map<String, Object> createPoint(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @RequestBody CreateTripPointRequest request) {
    return tripsService.createPoint(currentUserResolver.resolve(userId), tripId, request);
  }

  @PatchMapping("/{tripId}/points/{pointId}")
  public Map<String, Object> updatePoint(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @PathVariable String pointId,
      @RequestBody(required = false) CreateTripPointRequest request) {
    return tripsService.updatePoint(currentUserResolver.resolve(userId), tripId, pointId, request);
  }

  @DeleteMapping("/{tripId}/points/{pointId}")
  public Map<String, Object> deletePoint(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @PathVariable String pointId) {
    return tripsService.deletePoint(currentUserResolver.resolve(userId), tripId, pointId);
  }

  @PostMapping("/{tripId}/points/reorder")
  public Map<String, Object> reorderPoints(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @RequestBody ReorderTripPointsRequest request) {
    return tripsService.reorderPoints(currentUserResolver.resolve(userId), tripId, request);
  }

  @PostMapping("/{tripId}/auto-assemble")
  public Map<String, Object> autoAssemble(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @RequestBody(required = false) AutoAssembleTripRequest request) {
    return tripsService.autoAssembleTrip(currentUserResolver.resolve(userId), tripId, request);
  }

  @PostMapping("/{tripId}/publish")
  public Map<String, Object> publish(
      @RequestHeader(value = "x-user-id", required = false) String userId,
      @PathVariable String tripId,
      @RequestBody(required = false) PublishTripRequest request) {
    return tripsService.publishTrip(currentUserResolver.resolve(userId), tripId, request);
  }
}
